### PR TITLE
improve life cycle hooks typings

### PR DESCRIPTION
**Pull request by jantimon**
## Description

I would like to prepare a PR for #8643.  
Right now the life cycle hooks are using generics and reducers which kind of blocks an easy implementation.

Therefore I removed the reducer and the generic type.

Related #8643

## Type of change

Please delete options that are not relevant.

- [x] Internal Typings

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/utils/plugins-helpers/src/types.ts [hunk 1]
```diff
@@ -524,58 +524,60 @@ export namespace Types {
...
   export type HookFunction = (...args: any[]) => void | Promise<void>;
 
+  export type LifeCycleHookValue = string | HookFunction | (string | HookFunction)[];
+
   /**
    * @description All available lifecycle hooks
    * @additionalProperties false
    */
-  export type LifecycleHooksDefinition<T = string | HookFunction | (string | HookFunction)[]> = {
+  export type LifecycleHooksDefinition = {
```
Comment (saihaj) [inline packages/utils/plugins-helpers/src/types.ts:533]: this I guess would be a breaking type change 🤔

Comment (jantimon) [reply packages/utils/plugins-helpers/src/types.ts:533]: good point 👍
are those types exported?

should I adjust the changeset?

Comment (jantimon) [reply packages/utils/plugins-helpers/src/types.ts:533]: if I read https://unpkg.com/browse/@graphql-codegen/cli@2.14.1/typings/index.d.ts correctly `hooks.ts` is not exported.

So I guess it is not a breaking change - or did I miss sth?

Comment (saihaj) [reply packages/utils/plugins-helpers/src/types.ts:533]: Not in Cli but for the `@graphql-codegen/plugin-helpers` which I think many plugins use https://unpkg.com/browse/@graphql-codegen/plugin-helpers@2.7.2/typings/types.d.ts

Comment (jantimon) [reply packages/utils/plugins-helpers/src/types.ts:533]: Yes that's absolutely valid 👍 
I believe the risk that a plugin makes use of those generics is very low but it is definitely possible.

Should I change the changeset to be a breaking change?

File: .changeset/plenty-bottles-prove.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@graphql-codegen/cli': patch
+'@graphql-codegen/plugin-helpers': patch
```
Comment (saihaj) [inline .changeset/plenty-bottles-prove.md:3]:
```suggestion
'@graphql-codegen/plugin-helpers': major
```

File: packages/graphql-codegen-cli/src/hooks.ts [hunk 3]
```diff
@@ -15,37 +15,6 @@ const DEFAULT_HOOKS: Types.LifecycleHooksDefinition<string[]> = {
...
-    }
-    if (Array.isArray(_hooks[hookName])) {
-      return {
-        ...prev,
-        [hookName]: _hooks[hookName] as string[],
-      };
-    }
-    return prev;
-  }, {} as Types.LifecycleHooksDefinition<(string | Types.HookFunction)[]>);
-}
```
Comment (saihaj) [inline packages/graphql-codegen-cli/src/hooks.ts:47]: let's refactor in seperate PR and keep this just isolated for type changes

**Review by saihaj — CHANGES_REQUESTED** [review]

File: packages/graphql-codegen-cli/src/hooks.ts (hunk 3 above)
Comment (jantimon) [reply packages/graphql-codegen-cli/src/hooks.ts]: this code contains some type hacks (e.g. casting functions to strings) so for me it felt better to shop both changes at once.

but if it makes the review easier it's fine for me: https://github.com/dotansimha/graphql-code-generator/pull/8652/commits/b1cb681008d1a8ba15f61efed0c3fe639b6a1dda

**Review by saihaj — APPROVED** [review]

**Comment by saihaj** [comment]
sorry for the delay here

**Pull request merged**